### PR TITLE
Feature/implement skdataset caching

### PR DESCRIPTION
**Pull request by timesler**
<summary of the change, relevant motivation, and context>

Change type:

- [ ] Bug fix
- [x] Feature
- [x] Documentation

Checklist:

- [x] My code follows the style guidelines of this [project](https://docs.google.com/document/d/1jckZJe0CrWyF-IjxoO2OfBKAyKLC3Yk9Xr7UmOLBETA/edit?usp=sharing)
- [x] I have performed a self-review of my own code
- [x] I have commented my code and added docstrings to all exposed functions and classes
- [x] I have made corresponding changes to the documentation
- [x] Any relevant changes to third party licenses have been updated in the README

### CODE REVIEW

**Review by AngranLi — DISMISSED** [review]

File: xt_training/utils/sklearn_wrapper.py [hunk 1]
```diff
@@ -62,6 +69,82 @@ def __len__(self):
...
+    """
+
+    def __init__(
+        self, dataset, shuffle=False, num_workers=0, collate_fn=None, pin_memory=False, timeout=0,
+        worker_init_fn=None, multiprocessing_context=None, cache=True
+    ):
+        """Contructor for SKDataLoader class.
+
+        Args:
+            cache (bool, optional): Whether to store a cached version of the dataset to use for
```
Comment (simonrmonk) [inline xt_training/utils/sklearn_wrapper.py:114]: One very minor comment here: I think documenting the type of the argument here is no longer needed since argument annotations were introduced. So in this case, you could move `bool` into the init argument declaration as `cache: bool = True`. That way we've codified this comment and they can't get out of sync. It also gives us the added bonus of type hinting for users of this class. I also thing the word `optional` can be removed because having a default argument already indicates that the argument doesn't need to be explicitly provided by the user.

**Review by simonrmonk — DISMISSED** [review]
Hi @timesler, this looks good to me and all makes sense. I'm wondering if one day it would make sense to define more general interfaces for data loading as well as model training. We've added wrapper functionality to get sklearn models working with the pytorch way of doing things - which I think causes us to have to do some slightly awkward things and will likely make code maintenance more challenging. I think we could simplify things if we created more general interfaces and then wrapped both Pytorch and Sklearn. That way we could load data and train models using both frameworks without them having to have any awareness of how the other framework is doing things. Using OOP terminology, we would be encapsulating the parts that vary. I think this would have several advantages including (non-exhaustive list):
* simplified device management
* no need to force sklearn to imitate Pytorch (e.g. batch sizes, eval vs. train mode, etc.)
* easier to add another framework in the future

Maybe this is worth a meeting to discuss at some point.

File: xt_training/utils/sklearn_wrapper.py (hunk 1 above)
Comment (AngranLi) [reply xt_training/utils/sklearn_wrapper.py:114]: I have a question: if the variable can be more than one type, e.g. `int` or `list of int`, can we declare it in the argument declaration?

Comment (simonrmonk) [reply xt_training/utils/sklearn_wrapper.py:114]: Good question. Yes, you can using the `typing` library. It can be done like this:

```
from typing import Union, List

def func(arg: Union[int, List[int]]):
    ...
```
